File: Tarea13 (MergeSort)/Tarea13.java
public class Tarea13 {
    void merge(int[] a, int l, int m, int r) {
        int a1 = m - l + 1;
        int a2 = r - m;

        int[] L = new int[a1];
        int[] R = new int[a2];

        for (int i = 0; i < a1; i++)
            L[i] = a[l + i];
        for (int j = 0; j < a2; j++)
            R[j] = a[m + 1 + j];

        int i = 0, j = 0, k = l;

        while (i < a1 && j < a2) {
            if (L[i] <= R[j])
                a[k++] = L[i++];
            else
                a[k++] = R[j++];
        }

        while (i < a1) a[k++] = L[i++];
        while (j < a2) a[k++] = R[j++];
    }

    void mergeSort(int[] a, int l, int r) {
        if (l < r) {
            int m = l + (r - l) / 2;
            mergeSort(a, l, m);
            mergeSort(a, m + 1, r);
            merge(a, l, m, r);
        }
    }

    public static void main(String[] args) {
        int[] a = {39, 28, 44, 11};
        Tarea13 obj = new Tarea13();
        System.out.println("Antes de ordenar el arreglo:");
        for (int x : a) System.out.print(x + " ");
        obj.mergeSort(a, 0, a.length - 1);
        System.out.println("\nDespués de ordenar el arreglo:");
        for (int x : a) System.out.print(x + " ");
    }
}

File: Tarea13 (MergeSort)/Tarea13.py
def merge(a, l, m, r):
    a1 = m - l + 1  
    a2 = r - m      

    L = [0] * a1
    R = [0] * a2

    for j in range(a1):
        L[j] = a[l + j]
    for k in range(a2):
        R[k] = a[m + 1 + k]

    i = 0  
    j = 0 
    k = l 
    while i < a1 and j < a2:
        if L[i] <= R[j]:
            a[k] = L[i]
            i += 1
        else:
            a[k] = R[j]
            j += 1
        k += 1
    while i < a1:
        a[k] = L[i]
        i += 1
        k += 1
    while j < a2:
        a[k] = R[j]
        j += 1
        k += 1


def mergeSort(a, l, r):
    if l < r:
        m = l + (r - l) // 2
        mergeSort(a, l, m)
        mergeSort(a, m + 1, r)
        merge(a, l, m, r)


a = [39, 28, 44, 11]
s = len(a)
print("Antes de ordenar el arreglo:")
for x in a:
    print(x, end=" ")
mergeSort(a, 0, s - 1)
print("\nDespués de ordenar el arreglo:")
for x in a:
    print(x, end=" ")
